sdk/core/azure-core/src/main/java/com/azure/core/http/policy/HttpLoggingPolicy.java

```diff
@@ -52,105 +57,142 @@ public HttpLoggingPolicy(HttpLogOptions httpLogOptions) {
-                            bodyString = prettyPrintIfNeeded(
-                                logger,
-                                request.getHeaders().getValue("Content-Type"),
-                                bodyString);
-                            logger.info("{}-byte body:%n{}", contentLength, bodyString);
-                            logger.info("--> END {}", request.getHttpMethod());
-                            return Mono.empty();
-                        });
-                    } catch (Exception e) {
-                        reqBodyLoggingMono = Mono.error(e);
-                    }
+                String contentType = request.getHeaders().getValue("Content-Type");
+                long contentLength = getContentLength(logger, request.getHeaders());
+
+                if (bodyIsPrintable(contentType, contentLength)) {
+                    requestLoggingMono = FluxUtil.collectBytesInByteBufferStream(request.getBody()).flatMap(bytes ->
+                        Mono.just(requestLogMessage.append(String.format("%d-byte body:%n%s%n--> END %s%n",
+                            contentLength, new String(bytes, StandardCharsets.UTF_8), request.getHttpMethod()))
+                            .toString()));
                 } else {
-                    logger.info("{}-byte body: (content not logged)", contentLength);
-                    logger.info("--> END {}", request.getHttpMethod());
+                    requestLogMessage.append(String.format("%d-byte body: (content not logged)%n--> END %s%n",
+                        contentLength, request.getHttpMethod()));
                 }
             }
         }
-        return reqBodyLoggingMono;
+
+        return requestLoggingMono.doOnNext(logger::info);
     }
 
-    private void formatAllowableHeaders(Set<String> allowedHeaderNames, HttpHeaders requestResponseHeaders,
-        ClientLogger logger) {
-        Set<String> lowerCasedAllowedHeaderNames = allowedHeaderNames.stream().map(String::toLowerCase)
-            .collect(Collectors.toSet());
-        StringBuilder sb = new StringBuilder();
-        for (HttpHeader header : requestResponseHeaders) {
-            String headerName = header.getName();
-            sb.append(headerName).append(":");
-            if (lowerCasedAllowedHeaderNames.contains(headerName.toLowerCase(Locale.ROOT))) {
-                sb.append(header.getValue());
+    private Mono<HttpResponse> logResponse(final ClientLogger logger, final HttpResponse response, long startNs) {
+        /*
+         * Logging is either disabled or the logging level is above information (warning or error), this will result
+         * in nothing being logged so perform a no-op.
+         */
+        int numericLogLevel = LoggingUtil.getEnvironmentLoggingLevel().toNumeric();
+        if (numericLogLevel == LogLevel.DISABLED.toNumeric()
+            || numericLogLevel > LogLevel.INFORMATIONAL.toNumeric()) {
+            return Mono.just(response);
+        }
```

Comment:
This is common code between logRequest() and logResponse(). This can be extracted as a method to reduce duplication.

Reply:
Created a method for these.